### ISSUE TITLE
Constraint based test failure on CI

### ISSUE THREAD

**Issue by lehins** [IntersectMBO/cardano-ledger#4137]
There was a [test failure on CI](https://ci.iog.io/build/3058930) and here is a reproducer for current master a807afb1ae466775eceab099158c03ba861f5404:
```
    Constraint soundness:                                             FAIL (0.21s)
      *** Failed! Timeout of 100000 microseconds exceeded. (after 19 tests):
      OrderInfo {sumBeforeParts = False, sizeBeforeArg = False, setBeforeSubset = False}
      (,GenEnv {gOrder = OrderInfo {sumBeforeParts = False, sizeBeforeArg = False, setBeforeSubset = False}, gEnv = A -> Set{▵₳ -887 | size = 29}
      AA -> ▵₳ -68354
      AB -> Map{542 -> ₳ 11 | size = 9, sum = ₳ 86}
      AD -> Map{67 -> 366 | size = 8, sum = 39822}
      AE -> Map{▵₳ -797 -> ▵₳ -740 | size = 13}
      AF -> Map{542 -> ₳ 11 | size = 9, sum = ₳ 86}
      AG -> Map{98 -> ▵₳ -1 | size = 14}
      AI -> Set{₳ 6 | size = 6, sum = ₳ 94}
      AJ -> (Range 6 6)
      AN -> Map{139 -> 9603 | size = 15, sum = 82185}
      AO -> Map{3805 -> ▵₳ -22 | size = 5}
      B -> Map{102 -> ▵₳ 304 | size = 4}
      C -> Map{2527 -> ▵₳ 441 | size = 5}
      D -> Map{₳ 12 -> ₳ 7 | size = 4, sum = ₳ 31}
      E -> ₳ 158
      F -> Set{▵₳ -972 | size = 5}
      G -> Map{13 -> ▵₳ 232 | size = 13}
      H -> ▵₳ 887320
      I -> Map{▵₳ -887 -> ₳ 16 | size = 18, sum = ₳ 149}
      J -> Set{₳ 6 | size = 6, sum = ₳ 94}
      K -> [₳ 11 | size = 11, sum = Coin 121]
      L -> Set[98,1010,3935,4073,5641,7249,9938]
      M -> Map{₳ 1 -> ▵₳ 8 | size = 9}
      N -> Map{946 -> ▵₳ 775 | size = 4}
      O -> (Range 15 15)
      P -> Map{₳ 1 -> ▵₳ -446 | size = 13}
      Q -> (Range 5 5)
      R -> Map{43 -> 5718 | size = 15, sum = 68815}
      S -> ₳ 117
      T -> Map{289 -> ▵₳ -191 | size = 7}
      U -> (Range 5 5)
      V -> Map{2612 -> ▵₳ -358 | size = 13}
      W -> ₳ 117
      X -> Map{8435 -> ▵₳ 59 | size = 3}
      Y -> (Range 7 7)
      Z -> (Range 8 8)
      , gSolved = fromList [(A,3),(AA,2),(AB,1),(AE,2),(AF,2),(AI,3),(AJ,1),(B,1),(D,1),(E,1),(F,1),(H,1),(I,2),(J,2),(K,1),(M,4),(N,1),(O,1),(P,1),(Q,2),(S,1),(T,1),(U,2),(V,2),(W,2),(Y,1),(Z,1)]})
      Sized (Range 5 5) F
      SumsTo (Right (DeltaCoin (-2))) H > ∑ sum Map{1590 -> ▵₳ -449 | size = 3} + sum Map{165 -> ▵₳ 882 | size = 13}
      Sized (Range 4 4) (Dom D)
      (Dom Map{2527 -> ▵₳ 257 | size = 5}) :=: (Dom C)
      Set[13,447,1194,1463,2586,2954,3170,5679,6377,6699,6784,7351,7867] :=: (Dom G)
      (Dom Map{98 -> ₳ 17 | size = 7, sum = ₳ 51}) :=: L
      Disjoint Set{▵₳ -972 | size = 5} (Rng B)
      Sized (Range 13 13) (Dom P)
      Disjoint F (Dom I)
      Sized (Range 13 13) (Dom P)
      Sized Q F
      SumsTo (Right (Coin 0)) S <= ∑ sum Map{1940 -> ₳ 10 | size = 10, sum = ₳ 84} + sum Map{4503 -> ₳ 7 | size = 3, sum = ₳ 33}
      [₳ 11 | size = 11, sum = Coin 121] :=: K
      F ⊆  (Rng V)
      Disjoint (Dom D) J
      Sized O (Rng Map{1337 -> ▵₳ 177 | size = 15})
      Map{946 -> ▵₳ 775 | size = 4} :=: N
      SumsTo (Right (Coin 2)) E >= ∑ sum Map{377 -> ₳ 5 | size = 18, sum = ₳ 158}
      Sized Y Set{▵₳ -952 | size = 7}
      (Dom Map{438 -> ▵₳ -817 | size = 15}) :=: (Rng R)
      Sized Z (Rng Map{2011 -> ₳ 12 | size = 8, sum = ₳ 115})
      Sized U F
      F ⊆  (Rng V)
      (Dom Map{366 -> ▵₳ -793 | size = 8}) :=: (Rng AD)
      S :=: W
      Disjoint (Rng B) (Dom AE)
      (Dom I) ⊆  A
      Set[98,1010,1725,2544,2603,3935,4073,5377,5641,6661,7249,7795,7981,9938] :=: (Dom AG)
      SumsTo (Right (DeltaCoin 0)) AA <= ∑ sum Map{41 -> ▵₳ -259 | size = 12} + sum Map{3439 -> ▵₳ 655 | size = 1} + sum N
      Set{▵₳ -973 | size = 7} :=: (Rng T)
      Set[8435,9747,9919] :=: (Dom X)
      J :=: AI
      (Dom Map{365 -> ₳ 3 | size = 15, sum = ₳ 136}) :=: (Rng AN)
      Disjoint (empty::Set Coin) (Rng AB)
      AB :=: AF
      Sized (Range 29 29) A
      AI ⊆  (Dom M)
      Sized AJ (Dom Map{₳ 6 -> ▵₳ 461 | size = 6})
      (Dom Map{3805 -> ₳ 17 | size = 5, sum = ₳ 57}) :=: (Dom AO)
      
      
      -- Order --
      OrderInfo {sumBeforeParts = False, sizeBeforeArg = False, setBeforeSubset = False}
      
      -- Constraints --
      Sized (Range 5 5) F
      SumsTo (Right (DeltaCoin (-2))) H > ∑ sum Map{1590 -> ▵₳ -449 | size = 3} + sum Map{165 -> ▵₳ 882 | size = 13}
      Sized (Range 4 4) (Dom D)
      (Dom {2527 -> ▵₳ 257, 6581 -> ▵₳ 575, 9069 -> ▵₳ -654, 9401 -> ▵₳ -804, 9668 -> ▵₳ 34}) == (Dom C)
      {13, 447, 1194, 1463, 2586, 2954, 3170, 5679, 6377, 6699, 6784, 7351, 7867} == (Dom G)
      (Dom {98 -> ₳ 17, 1010 -> ₳ 1, 3935 -> ₳ 12, 4073 -> ₳ 3, 5641 -> ₳ 1, 7249 -> ₳ 6, 9938 -> ₳ 11}) == L
      Disjoint {▵₳ -972, ▵₳ -783, ▵₳ -386, ▵₳ -358, ▵₳ 608} (Rng B)
      Sized (Range 13 13) (Dom P)
      Disjoint F (Dom I)
      Sized (Range 13 13) (Dom P)
      Sized Q F
      SumsTo (Right (Coin 0)) S <= ∑ sum Map{1940 -> ₳ 10 | size = 10, sum = ₳ 84} + sum Map{4503 -> ₳ 7 | size = 3, sum = ₳ 33}
      [₳ 11 | size = 11, sum = Coin 121] == K
      F ⊆ (Rng V)
      Disjoint (Dom D) J
      Sized O (Rng Map{1337 -> ▵₳ 177 | size = 15})
      {946 -> ▵₳ 775, 2565 -> ▵₳ -298, 2714 -> ▵₳ -284, 5543 -> ▵₳ -719} == N
      SumsTo (Right (Coin 2)) E >= ∑ sum Map{377 -> ₳ 5 | size = 18, sum = ₳ 158}
      Sized Y Set{▵₳ -952 | size = 7}
      (Dom {438 -> ▵₳ -817, 616 -> ▵₳ 820, 1236 -> ▵₳ -214, 3788 -> ▵₳ 831, 3965 -> ▵₳ 267, 4290 -> ▵₳ -914, 4697 -> ▵₳ -521, 5153 -> ▵₳ -752, 5216 -> ▵₳ 890, 5710 -> ▵₳ -380, 5718 -> ▵₳ -993, 6594 -> ▵₳ -963, 6834 -> ▵₳ 218, 6923 -> ▵₳ 548, 7637 -> ▵₳ -357}) == (Rng R)
      Sized Z (Rng Map{2011 -> ₳ 12 | size = 8, sum = ₳ 115})
      Sized U F
      F ⊆ (Rng V)
      (Dom {366 -> ▵₳ -793, 2012 -> ▵₳ 743, 3563 -> ▵₳ 264, 4328 -> ▵₳ -452, 5218 -> ▵₳ -689, 6639 -> ▵₳ -468, 8348 -> ▵₳ 637, 9348 -> ▵₳ -549}) == (Rng AD)
      S == W
      Disjoint (Rng B) (Dom AE)
      (Dom I) ⊆ A
      {98, 1010, 1725, 2544, 2603, 3935, 4073, 5377, 5641, 6661, 7249, 7795, 7981, 9938} == (Dom AG)
      SumsTo (Right (DeltaCoin 0)) AA <= ∑ sum Map{41 -> ▵₳ -259 | size = 12} + sum Map{3439 -> ▵₳ 655 | size = 1} + sum N
      {▵₳ -973, ▵₳ -554, ▵₳ -366, ▵₳ -248, ▵₳ -191, ▵₳ -3, ▵₳ 851} == (Rng T)
      {8435, 9747, 9919} == (Dom X)
      J == AI
      (Dom {365 -> ₳ 3, 1381 -> ₳ 12, 1468 -> ₳ 5, 2842 -> ₳ 10, 3077 -> ₳ 7, 4155 -> ₳ 2, 4893 -> ₳ 15, 5003 -> ₳ 9, 6694 -> ₳ 12, 7431 -> ₳ 14, 7495 -> ₳ 11, 8589 -> ₳ 12, 9402 -> ₳ 4, 9603 -> ₳ 12, 9787 -> ₳ 8}) == (Rng AN)
      Disjoint {} (Rng AB)
      AB == AF
      Sized (Range 29 29) A
      AI ⊆ (Dom M)
      Sized AJ (Dom Map{₳ 6 -> ▵₳ 461 | size = 6})
      (Dom {3805 -> ₳ 17, 8240 -> ₳ 5, 8517 -> ₳ 15, 9443 -> ₳ 6, 9947 -> ₳ 14}) == (Dom AO)
      
      -- Model solution --
      A :: Set DeltaCoin -> {▵₳ -887, ▵₳ -832, ▵₳ -778, ▵₳ -769, ▵₳ -631, ▵₳ -574, ▵₳ -503, ▵₳ -470, ▵₳ -424, ▵₳ -350, ▵₳ -310, ▵₳ -279, ▵₳ -256, ▵₳ -205, ▵₳ -173, ▵₳ -102, ▵₳ -92, ▵₳ -74, ▵₳ -71, ▵₳ -61, ▵₳ 2, ▵₳ 214, ▵₳ 361, ▵₳ 462, ▵₳ 592, ▵₳ 663, ▵₳ 699, ▵₳ 829, ▵₳ 872}
      AA :: DeltaCoin -> ▵₳ -68354
      AB :: Map Int Coin -> {542 -> ₳ 11, 1003 -> ₳ 8, 5982 -> ₳ 7, 6453 -> ₳ 14, 7159 -> ₳ 10, 7754 -> ₳ 5, 9170 -> ₳ 13, 9624 -> ₳ 17, 9999 -> ₳ 1}
      AD :: Map Int Int -> {67 -> 366, 1153 -> 6639, 2226 -> 5218, 6039 -> 8348, 6548 -> 4328, 6855 -> 9348, 7430 -> 3563, 7580 -> 2012}
      AE :: Map DeltaCoin DeltaCoin -> {▵₳ -797 -> ▵₳ -740, ▵₳ -732 -> ▵₳ -854, ▵₳ -597 -> ▵₳ 536, ▵₳ -585 -> ▵₳ -202, ▵₳ -580 -> ▵₳ 783, ▵₳ -309 -> ▵₳ 383, ▵₳ -220 -> ▵₳ 501, ▵₳ 71 -> ▵₳ -558, ▵₳ 197 -> ▵₳ 962, ▵₳ 544 -> ▵₳ -107, ▵₳ 594 -> ▵₳ -700, ▵₳ 792 -> ▵₳ 834, ▵₳ 991 -> ▵₳ 310}
      AF :: Map Int Coin -> {542 -> ₳ 11, 1003 -> ₳ 8, 5982 -> ₳ 7, 6453 -> ₳ 14, 7159 -> ₳ 10, 7754 -> ₳ 5, 9170 -> ₳ 13, 9624 -> ₳ 17, 9999 -> ₳ 1}
      AG :: Map Int DeltaCoin -> {98 -> ▵₳ -1, 1010 -> ▵₳ 889, 1725 -> ▵₳ -366, 2544 -> ▵₳ -197, 2603 -> ▵₳ 927, 3935 -> ▵₳ 114, 4073 -> ▵₳ -897, 5377 -> ▵₳ 672, 5641 -> ▵₳ 213, 6661 -> ▵₳ -302, 7249 -> ▵₳ 690, 7795 -> ▵₳ -54, 7981 -> ▵₳ 872, 9938 -> ▵₳ 216}
      AI :: Set Coin -> {₳ 6, ₳ 7, ₳ 15, ₳ 17, ₳ 18, ₳ 31}
      AJ :: Size -> (Range 6 6)
      AN :: Map Int Int -> {139 -> 9603, 271 -> 2842, 384 -> 1381, 1144 -> 4893, 1302 -> 9787, 2090 -> 365, 3836 -> 7495, 5213 -> 3077, 6355 -> 8589, 6770 -> 6694, 6776 -> 5003, 7979 -> 7431, 8132 -> 4155, 8914 -> 9402, 9706 -> 1468}
      AO :: Map Int DeltaCoin -> {3805 -> ▵₳ -22, 8240 -> ▵₳ 849, 8517 -> ▵₳ 596, 9443 -> ▵₳ -25, 9947 -> ▵₳ 583}
      B :: Map Int DeltaCoin -> {102 -> ▵₳ 304, 5422 -> ▵₳ 353, 6173 -> ▵₳ 983, 6589 -> ▵₳ -848}
      C :: Map Int DeltaCoin -> {2527 -> ▵₳ 441, 6581 -> ▵₳ 900, 9069 -> ▵₳ 143, 9401 -> ▵₳ 170, 9668 -> ▵₳ 431}
      D :: Map Coin Coin -> {₳ 12 -> ₳ 7, ₳ 14 -> ₳ 10, ₳ 16 -> ₳ 10, ₳ 25 -> ₳ 4}
      E :: Coin -> ₳ 158
      F :: Set DeltaCoin -> {▵₳ -972, ▵₳ -783, ▵₳ -386, ▵₳ -358, ▵₳ 608}
      G :: Map Int DeltaCoin -> {13 -> ▵₳ 232, 447 -> ▵₳ 593, 1194 -> ▵₳ -711, 1463 -> ▵₳ -761, 2586 -> ▵₳ 381, 2954 -> ▵₳ 436, 3170 -> ▵₳ 266, 5679 -> ▵₳ 689, 6377 -> ▵₳ -576, 6699 -> ▵₳ 398, 6784 -> ▵₳ 420, 7351 -> ▵₳ 337, 7867 -> ▵₳ 977}
      H :: DeltaCoin -> ▵₳ 887320
      I :: Map DeltaCoin Coin -> {▵₳ -887 -> ₳ 16, ▵₳ -832 -> ₳ 12, ▵₳ -778 -> ₳ 11, ▵₳ -631 -> ₳ 8, ▵₳ -574 -> ₳ 4, ▵₳ -503 -> ₳ 17, ▵₳ -350 -> ₳ 8, ▵₳ -256 -> ₳ 6, ▵₳ -205 -> ₳ 8, ▵₳ -102 -> ₳ 1, ▵₳ -74 -> ₳ 13, ▵₳ -71 -> ₳ 6, ▵₳ -61 -> ₳ 10, ▵₳ 2 -> ₳ 7, ▵₳ 663 -> ₳ 3, ▵₳ 699 -> ₳ 9, ▵₳ 829 -> ₳ 6, ▵₳ 872 -> ₳ 4}
      J :: Set Coin -> {₳ 6, ₳ 7, ₳ 15, ₳ 17, ₳ 18, ₳ 31}
      K :: [Coin] -> [₳ 11 | size = 11, sum = Coin 121]
      L :: Set Int -> {98, 1010, 3935, 4073, 5641, 7249, 9938}
      M :: Map Coin DeltaCoin -> {₳ 1 -> ▵₳ 8, ₳ 2 -> ▵₳ -778, ₳ 5 -> ▵₳ 795, ₳ 6 -> ▵₳ -471, ₳ 7 -> ▵₳ 464, ₳ 15 -> ▵₳ -271, ₳ 17 -> ▵₳ -601, ₳ 18 -> ▵₳ -474, ₳ 31 -> ▵₳ -431}
      N :: Map Int DeltaCoin -> {946 -> ▵₳ 775, 2565 -> ▵₳ -298, 2714 -> ▵₳ -284, 5543 -> ▵₳ -719}
      O :: Size -> (Range 15 15)
      P :: Map Coin DeltaCoin -> {₳ 1 -> ▵₳ -446, ₳ 6 -> ▵₳ 392, ₳ 7 -> ▵₳ -606, ₳ 8 -> ▵₳ 270, ₳ 9 -> ▵₳ 711, ₳ 10 -> ▵₳ 405, ₳ 11 -> ▵₳ -464, ₳ 12 -> ▵₳ -47, ₳ 15 -> ▵₳ 192, ₳ 16 -> ▵₳ 339, ₳ 17 -> ▵₳ -417, ₳ 18 -> ▵₳ -607, ₳ 23 -> ▵₳ -173}
      Q :: Size -> (Range 5 5)
      R :: Map Int Int -> {43 -> 5718, 455 -> 4290, 1024 -> 438, 1395 -> 616, 1503 -> 7637, 2047 -> 6594, 2808 -> 6923, 2830 -> 6834, 4546 -> 3788, 5189 -> 3965, 5348 -> 5216, 5751 -> 1236, 6199 -> 5153, 6998 -> 4697, 8221 -> 5710}
      S :: Coin -> ₳ 117
      T :: Map Int DeltaCoin -> {289 -> ▵₳ -191, 2420 -> ▵₳ -366, 2777 -> ▵₳ -248, 4107 -> ▵₳ -973, 4240 -> ▵₳ -3, 5193 -> ▵₳ -554, 9187 -> ▵₳ 851}
      U :: Size -> (Range 5 5)
      V :: Map Int DeltaCoin -> {2612 -> ▵₳ -358, 3380 -> ▵₳ -497, 3765 -> ▵₳ -972, 3914 -> ▵₳ -816, 4048 -> ▵₳ 871, 4619 -> ▵₳ -386, 5707 -> ▵₳ 928, 6536 -> ▵₳ 592, 7336 -> ▵₳ 689, 7441 -> ▵₳ -783, 7708 -> ▵₳ -520, 8512 -> ▵₳ 608, 8828 -> ▵₳ 370}
      W :: Coin -> ₳ 117
      X :: Map Int DeltaCoin -> {8435 -> ▵₳ 59, 9747 -> ▵₳ -859, 9919 -> ▵₳ 463}
      Y :: Size -> (Range 7 7)
      Z :: Size -> (Range 8 8)
      
      Use --quickcheck-replay=567654 to reproduce.
      Use -p '/Constraint soundness/' to rerun this test only.
```

**Comment by MaximilianAlgehed** [comment]
I think this is one for @TimSheard - it's in the old system and he is more familiar than I am with that code.

**Comment by lehins** [comment]
Fixed in #4139
